src/common/dao/project_blob.go

```diff
@@ -134,3 +134,45 @@ WHERE af.project_id = ?
 
 	return size, err
 }
+
+// RemoveUntaggedBlobs ...
+func RemoveUntaggedBlobs(pid int64) error {
+	var projectBlobs []*models.ProjectBlob
+
+	sql := `
+SELECT 
+    *
+FROM project_blob AS pb
+WHERE project_id = ? 
+AND pb.blob_id NOT IN (
+SELECT 
+    DISTINCT bb.id
+FROM artifact af
+JOIN artifact_blob afnb
+    ON af.digest = afnb.digest_af
+JOIN BLOB bb
+    ON afnb.digest_blob = bb.digest
+WHERE af.project_id = ?)
+`
+	_, err := GetOrmer().Raw(sql, pid, pid).QueryRows(&projectBlobs)
+	if err != nil {
+		return err
+	}
+
+	var pbIDs []interface{}
+	for _, pb := range projectBlobs {
+		pbIDs = append(pbIDs, pb.ID)
+	}
+
+	if len(pbIDs) == 0 {
+		return nil
+	}
+
+	sql = fmt.Sprintf(`DELETE FROM project_blob WHERE id IN (%s)`, ParamPlaceholderForIn(len(pbIDs)))
```

Comment:
The two sql can be merged into one

Reply:
as the pq will raise an exception when the IN or NOT IN condition array is empty.